src/http-util/src/proxy.rs

```diff
@@ -0,0 +1,366 @@
+        .unwrap_or(NoProxy::None);
+
+    ProxyConfig {
+        http_proxy,
+        https_proxy,
+        all_proxy,
+        no_proxy,
+    }
+}
+
+/// The analyzed proxy configuration.
+#[derive(Debug, Clone)]
+pub struct ProxyConfig {
+    http_proxy: Option<Uri>,
+    https_proxy: Option<Uri>,
+    all_proxy: Option<Uri>,
+    no_proxy: NoProxy,
+}
+
+impl ProxyConfig {
+    /// Returns the proxy to use for HTTP requests, if any.
+    pub fn http_proxy(&self) -> Option<&Uri> {
+        self.http_proxy.as_ref()
+    }
+
+    /// Returns the proxy to use for HTTPS requests, if any.
+    pub fn https_proxy(&self) -> Option<&Uri> {
+        self.https_proxy.as_ref()
+    }
+
+    /// Returns the proxy to use for HTTP or HTTPS requests, if any.
+    ///
+    /// The proxies returned by [`ProxyConfig::http_proxy`] and
+    /// [`ProxyConfig::https_proxy`] take precedence if present.
+    pub fn all_proxy(&self) -> Option<&Uri> {
+        self.all_proxy.as_ref()
+    }
+
+    /// Reports whether the request composed of the given URL scheme, host, and
+    /// port should be excluded from proxying.
+    pub fn exclude(&self, scheme: Option<&str>, host: Option<&str>, port: Option<u16>) -> bool {
+        self.no_proxy.matches(scheme, host, port)
+    }
+}
+
+#[derive(Debug, Clone)]
+enum NoProxy {
+    None,
+    Some {
+        ips: Vec<IpMatcher>,
+        hosts: Vec<DomainMatcher>,
+    },
+    All,
+}
+
+impl NoProxy {
+    fn parse(s: &str) -> NoProxy {
+        match s.trim() {
+            "" => NoProxy::None,
+            "*" => NoProxy::All,
+            _ => {
+                let mut ips = vec![];
+                let mut hosts = vec![];
+                for host in s.split(',') {
+                    let host = host.trim();
+
+                    // If the entire entry is a CIDR-formatted IP prefix,
+                    // we're done. To match Go, we don't allow these to carry
+                    // port constraints.
+                    if let Ok(net) = IpNet::from_str(host) {
+                        ips.push(IpMatcher { net, port: None });
+                        continue;
+                    }
+
+                    // See if we can split into a host and port. Ignore errors,
+                    // because bare IPv6 addresses are not valid authorities but
+                    // they *are* valid in `no_proxy`.
+                    let authority = Authority::from_str(host).ok();
+                    let (mut host, port) = match &authority {
+                        Some(authority) => (authority.host(), authority.port_u16()),
+                        None => (host, None),
+                    };
+
+                    // Trim the brackets surrounding an IPv6 address, if
+                    // present.
+                    if let Some(h) = host.strip_prefix('[') {
+                        if let Some(h) = h.strip_suffix(']') {
+                            host = h;
+                        }
+                    }
+
```

Comment:
Is whitespace inside the brackets, i.e. `[ 1:2:3::4   ]` allowed? If so, let's trim again here (unless `IpAddr::from_str` does it for us)

Reply:
It is not allowed!